### PR TITLE
fix MSVC 14.1 warnings-as-errors=on build

### PR DESCRIPTION
**Pull request by ajax16384**
fix type narrowing warns
use span random_bytes

### CODE REVIEW

File: src/receive_buffer.cpp [hunk 1]
```diff
@@ -125,7 +125,7 @@ void receive_buffer::cut(int const size, int const packet_size, int const offset
 		m_recv_end -= size;
 
 #if TORRENT_USE_ASSERTS
-		std::fill(m_recv_buffer.begin() + m_recv_end, m_recv_buffer.end(), 0xcc);
+		std::fill(m_recv_buffer.begin() + m_recv_end, m_recv_buffer.end(), static_cast<unsigned char>(0xcc));
```
Comment (arvidn) [inline src/receive_buffer.cpp]: as far as I can tell, ``m_recv_buffer.begin()`` returns a ``char*``, not ``unsigned char``. Am I missing something?

File: src/bt_peer_connection.cpp [hunk 2]
```diff
@@ -2084,7 +2084,7 @@ namespace libtorrent {
 
 		if (t->is_seed())
 		{
-			std::fill_n(ptr, packet_size - 5, 0xff);
+			std::fill_n(ptr, packet_size - 5, static_cast<std::uint8_t>(0xff));
```
Comment (arvidn) [inline src/bt_peer_connection.cpp]: if this works: ``std::uint8_t{0xff}``, I think I would prefer that syntax

File: src/bt_peer_connection.cpp [hunk 3]
```diff
@@ -2716,7 +2716,7 @@ namespace libtorrent {
 					disconnect(errors::no_memory, op_encryption);
 					return;
 				}
-				std::fill(m_sync_vc.get(), m_sync_vc.get() + 8, 0);
+				std::fill(m_sync_vc.get(), m_sync_vc.get() + 8, static_cast<char>(0));
```
Comment (arvidn) [inline src/bt_peer_connection.cpp]: here as well, if ``char{0}`` works

File: src/i2p_stream.cpp [hunk 4]
```diff
@@ -130,7 +131,7 @@ namespace libtorrent {
 		m_state = sam_connecting;
 
 		char tmp[20];
-		std::generate(tmp, tmp + sizeof(tmp), &std::rand);
+		aux::random_bytes(tmp);
```
Comment (arvidn) [inline src/i2p_stream.cpp]: +1

**Comment by arvidn** [comment]
lgtm (apart from comments).
also, if this actually eliminates all the warnings in the tests, please add the flag to the appveyor build command line as well!

File: src/torrent.cpp [hunk 5]
```diff
@@ -4865,14 +4865,11 @@ namespace libtorrent {
...
-
-		// initialize pad files to priority 0
-		for (file_index_t i(0); i < limit; ++i)
+		auto si = files.begin();
+		for (file_index_t i(0); i < limit; ++i, ++si)
 		{
-			if (!fs.pad_file_at(i)) continue;
-			m_file_priority[i] = 0;
+			// initialize pad files to priority 0
+			m_file_priority[i] = fs.pad_file_at(i) ? 0 : aux::numeric_cast<std::uint8_t>(*si);
```
Comment (aldenml) [inline src/torrent.cpp]: I suggest something like `aux::clamp(*si, 0, 7) & 0xff` here, what do you think?

Comment (arvidn) [inline src/torrent.cpp]: is ``& 0xff`` there to silence some warning? I don't see how ``clamp(..., 0, 7)`` could return anything outside of the ``0xff`` mask

Comment (arvidn) [inline src/torrent.cpp]: but yes, ``clamp()`` would have higher accuracy in fixing incorrect priority values. The distinction, though, that ``numeric_cast`` is only an assert and not a runtime check. Should the input vector be considered trusted? I'm not sure.

File: src/receive_buffer.cpp (hunk 1 above)
Comment (ajax16384) [inline src/receive_buffer.cpp]:
````c++
vc\tools\msvc\14.10.25017\include\xutility(2690): warning C4244: '=': conversion from 'const int' to 'char', possible loss of data
````

File: src/bt_peer_connection.cpp (hunk 2 above)
Comment (ajax16384) [inline src/bt_peer_connection.cpp]: ok

**Comment by ajax16384** [comment]
appveyor already has "warnings-as-errors=on"

**Comment by arvidn** [comment]
how come the appveyor build passes without this patch, if warnings are treated as errors and this patch fixes warnings?

**Comment by arvidn** [comment]
lgtm btw. I will merge steven's multi-home patch ahead of this though

**Comment by ajax16384** [comment]
patch fixes 14.1 warns (vs2017)

File: src/torrent.cpp (hunk 5 above)
Comment (aldenml) [inline src/torrent.cpp]: yes, `& 0xff` is a warning thing

**Pull request merged**